### ISSUE TITLE
when specific objects are selected, parent transforms are not applied

### ISSUE THREAD

**Issue by lexelby** [fablabnbg/inkscape-silhouette#43]
Let's say I have a document with red and blue lines.  I want to cut the red ones and score the blue ones.  I'll select a red line, do *Edit* -> *Select Same* -> *Stroke Color* to select all the red ones.  If some of the lines are in groups, all hell breaks loose.

The problem is that `recursivelyTraverseSvg` is being passed the selected paths directly, not the groups they're in.  If a line is in a group with a transform, that transform is not applied to the line.  Stuff ends up all over the place.

I'm not sure what the best solution is.  [inkscape-embroidery](https://www.github.com/lexelby/inkscape-embroidery) gathers transforms differently, using `simpletransform.composeParents()`.  Perhaps that would help?

**Comment by lexelby** [comment]
fixed in #45

**Comment by EtherGraf** [comment]
Just quickly tested your solution from #45.
I can confirm, it works well for me.

**Comment by EtherGraf** [comment]
@lexelby
Could you create a separate PR for this issue from your branch? Without the multiple cutting?
I will merge it into trunk then and close this issue.

Wish you a happy new year!

**Comment by jnweiger** [comment]
I am also interested in that fix. I've refactored the svg gatherer into a class of its own: 
https://github.com/jnweiger/inkscape-thunderlaser/blob/master/src/inksvg.py
same fix probably needed there too.

Hey everybody, thanks for your contributions, have great xmas and a happy 2018!

**Comment by jnweiger** [comment]
https://github.com/fablabnbg/inkscape-silhouette/pull/45/commits/9fe59ec5b1f4d53ec2f1b7159d3a5e43d5fce10c has most of it, right?

**Comment by jnweiger** [comment]
With class InkSvg there is a method recursivelyGetEnclosingTransform(self.selected[id]) which does the needed composeTransform() calls. 

I am using it like this:
```
if self.options.ids:
            # Traverse the selected objects
            for id in self.options.ids:
                transform = svg.recursivelyGetEnclosingTransform(self.selected[id])
                svg.recursivelyTraverseSvg([self.selected[id]], transform)
```
I cannot reproduce any falsely positioned lines with that code. 
So, how about refactoring the silhouette code to use inksvg.py ?

**Comment by EtherGraf** [comment]
> 9fe59ec has most of it, right?

Seems so, but two checkins later in d45d68ae2176ab94ac01c8bbae03356d9b4c66b5 the same code is touched again. Bugfix or extension?

**Comment by EtherGraf** [comment]
From looking at the sourcecode of inksvg, it seems containing the same bug. 

The difference is to "compose the transformations recursively" versus "calling composeParents() at each level"

**Comment by lexelby** [comment]
Whew... that fix was quite awhile ago in terms of my fork, so I'd have to go back in time and figure this all out again in order to create a separate PR.  I _think_ the two commits mentioned above are the whole fix, but I'm not certain.

I like @jnweiger's idea of switching over to inksvg better.  Then we can fix the problem once and be done with it.  At some point I'll probably port [inkscape-embroidery](https://github.com/lexelby/inkscape-embroidery) over to use inksvg as well.

**Comment by lexelby** [comment]
> From looking at the sourcecode of inksvg, it seems containing the same bug.
>
> The difference is to "compose the transformations recursively" versus "calling composeParents() at each level"

Agreed; inkSvg still has the same problem.  If the `aNodeList` argument to `recursivelyTraverseSvg()` contains objects that are inside groups, then the transforms of containing groups are not used.

To reproduce:

1. Draw two lines.
2. Group them together.
3. Select the group (not the two objects in it).
4. Rotate the group some amount.
5. Check with Inkscape's XML tool (ctrl-shift-x, I think) to ensure that the group has a transform property but the lines don't.

Now select just one of the lines inside the group.  This may be easiest to do if you use the Objects palette (new in inkscape 0.92). If you process that object with inkSvg, I think you'll find that the rotation is not applied properly.

**Comment by lexelby** [comment]
Oh!  I understand your comment now, @jnweiger.  Yes, I believe `recursivelyGetEnclosingTransform()` does the job.  I think you can probably simplify this using `simpletransform.composeParents()`, though?

**Comment by EtherGraf** [comment]
For a quick fix I have merged PR #56 now.
As a long term solution the switch to inksvg may be the better solution. I just don't know it.
Should I keep this issue open for further discusison?
